storage/innobase/dict/dict0dd.cc

```diff
@@ -1299,36 +1316,72 @@ int dd_table_open_on_dd_obj(dd::cache::Dictionary_client *client,
-  uint64 table_id = dd_table.se_private_id();
+std::tuple<int, std::vector<dict_table_t *>> dd_table_load_on_dd_obj(
+    dd::cache::Dictionary_client *client, space_id_t space_id,
+    const dd::Table &dd_table, table_id_t table_id, THD *thd,
+    const dd::String_type *schema_name, bool implicit) {
+  std::vector<dict_table_t *> table_vec;
 
-  if (table_id == dd::INVALID_OBJECT_ID) {
+  if (dd_table.se_private_id() == dd::INVALID_OBJECT_ID) {
     /* Partitioned table */
     ut_ad(dd_table_is_partitioned(dd_table));
+    auto end = dd_table.leaf_partitions().end();
+
+    if (table_id != 0) {
+      // Load specific partition based on table_id
+      auto i =
+          std::search_n(dd_table.leaf_partitions().begin(), end, 1, table_id,
+                        [](const dd::Partition *p, table_id_t id) {
+                          return (p->se_private_id() == id);
+                        });
+      if (i == end) {
+        // The partition table_id doesn't exist in this Table partitions.
+        return std::make_tuple(1, table_vec);
+      } else {
+        auto &part = *i;
 
-    for (auto part : dd_table.leaf_partitions()) {
-      uint64 tid = part->se_private_id();
-      int ret = dd_table_load_part(tid, dd_table, part, table, thd, schema_name,
-                                   implicit, space_id);
-      if (ret != 0) {
-        return ret;
+        uint64 tid = part->se_private_id();
+        dict_table_t *table = dd_table_load_part(
+            tid, dd_table, part, thd, schema_name, implicit, space_id);
+        if (table != nullptr) {
+          table_vec.push_back(table);
+        } else {
+          // Unable to load table. Error
+          return std::make_tuple(1, table_vec);
+        }
+      }
+    } else {
+      // Load all partitions
+      for (auto part : dd_table.leaf_partitions()) {
+        uint64 tid = part->se_private_id();
+
+        dict_table_t *table = dd_table_load_part(
+            tid, dd_table, part, thd, schema_name, implicit, space_id);
+        if (table != nullptr) {
+          table_vec.push_back(table);
+        } else {
+          // Unable to load table. Error
+          return std::make_tuple(1, table_vec);
+        }
       }
-      dd_table_close(table, thd, nullptr, false);
     }
-    return 0;
   } else {
+    // Load a NON-Partition table
+    if (table_id == 0) {
+      table_id = dd_table.se_private_id();
+    }
+
     const dd::Partition *dd_part = nullptr;
-    int ret = dd_table_load_part(table_id, dd_table, dd_part, table, thd,
-                                 schema_name, implicit, space_id);
-    if (ret != 0) {
-      return ret;
+    dict_table_t *table = dd_table_load_part(table_id, dd_table, dd_part, thd,
+                                             schema_name, implicit, space_id);
+    if (table != nullptr) {
+      table_vec.push_back(table);
+    } else {
+      // Unable to load table. Error
```

Comment:
Let me think. Lets say DDL is in progress, new indexes added, instant columns added etc.
1. ib_sdi_set() is called from DDL, which would delete mark current SDI record and a new SDI entry is added.
2. changes to mysql.ibd are also in same transaction.

Lets assume PXB copies a tablespace that is under DDL, we could have missed those un-flushed SDI changes but these changes are 'redo-logged' and undo-logged as well. 

The initial scan of mysql.ibd will be up to date physically because the redo phase is over. The only problem is DDL has to be rolledback, then mysql.ibd might have seen extra dd_index/column entries.

So if we do dictionary trx rollback before user trx rollback, we should be OK but this is a big grey area..

I am OK with adding debug assert for now.